Tests/RegexTests/RegexDSLTests.swift

```diff
@@ -559,11 +559,11 @@ class RegexDSLTests: XCTestCase {
 
   func testBackreference() throws {
     try _testDSLCaptures(
-      ("abc#41#42abc#42#42", ("abc#41#42abc#42#42", "abc", 42, "#42", nil)),
+      ("abc#41#42abcabcabc", ("abc#41#42abcabcabc", "abc", 42, "abc", nil)),
       matchType: (Substring, Substring, Int?, Substring?, Substring?).self, ==)
     {
-      let a = Reference()
-      let b = Reference()
+      let a = Reference(Substring.self)
+      let b = Reference(Int.self)
```

Comment:
Do we have tests where the types don't match up? What happens?

Reply:
You won't be able to capture something onto a `Reference` of a non-matching type. That's a type error.

Reply:
That's very compelling and alleviates the need for another way to name captures.

BTW, outside of scope of this PR, are we still trying to figure out what happens when a regex literal has names? Do we drop them completely, synthesize these references, or also have a stringy lookup?

Reply:
I think we drop them completely with the `._Unlabeld` extension I'm working on.  I'm aiming to create a PR for that later today.